marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -767,9 +767,9 @@ getUtxoFromTxOut
   -> C.TxIn -- ^ unique id and position of this transaction
   -> C.TxOut C.CtxTx era -- ^ Cardano TxOut
   -> Maybe Utxo -- ^ Utxo
-getUtxoFromTxOut maybeTargetAddresses (C.TxIn txid txix) (C.TxOut addr val dtum refScript) =
-  if isAddressInTarget maybeTargetAddresses addrAny
-  then Just $ Utxo
+getUtxoFromTxOut maybeTargetAddresses (C.TxIn txid txix) (C.TxOut addr val dtum refScript) = do
+  guard $ isAddressInTarget maybeTargetAddresses addrAny
```

Comment:
I didn't plan to generalise it, I just take advantage of `Maybe` being an alternative.

Reply:
But to whom this is an advantage?

Reply:
You avoid the introduction of a control structure, and thus reduce the complexity of the code and remove noise.
Here, it doesn't matter that much as we only have one condition and a few line of code, but in the generalise case it's a best practice that avoid nested block of code.

Reply:
if-else like the do-notation are both syntactic sugar, and there is no way around looking at the result of what `isAddressInTarget` returns, so it's not like that control structure is going away. Similarly, every line of a do-notation is a nested block of code because it's a function that takes the result of the previous line as input.

Given that the type signature of the function still returns a Maybe, then this change looks like replacing a straight implementation with with an indirect one that goes through Maybe's Alternative implementation. It also adds `do` which I guess means a Monad instance is also required. All these instances happen to exist for Maybe, but does this really remove noise and complexity from the code?

(But when this function really is an Alternative, then why not convince yourself that it is and then change the type signature to return `Alternative f => ... -> f Utxo`?)

Reply:
It's not about computational complexity, it's about readability: explicit nested control structures have a heavier cognitive burden because you need to remember in which branch you have. It's not a big deal when you have two branches, it may become the case when you start adding some. I think it's a good practice to remove them, even when the current complexity is bearable.
I'm totally fine with the `Alternative` type signature but I'm reluctant to signature changes, even for a more general one, unless we do think it provides added value. If you think it's the case, why not.

Reply:
But this change adds cognitive burden because first I need to know what Alternative does to Maybe. And second, I'd still need to convince myself that the function as a whole is an Alternative (that it will keep working when you add more code to it), but I personally don't know that. Otherwise it's just an accident that the Alternative version does the same thing as the direct version.

Reply:
I honesty took `guard` as known. Mainly because it removes complexity (to me at least) in such cases.
I guess we could discuss in a technical meeting whether we want to include such functions in our codebase or not (and extend this discussion to a broader set of functions).
In the meantime, as it isn't required, we can remove it, but I still think the change would have been valuable.

Reply:
But it's not about being allowed to use `guard`, it's just that (for example) `if cond then Just 0 else Nothing` is (IMO) easier to understand than `guard condition; pure empty` where the value depends on the particular type and the type classes being used[1], and if the return type isn't polymorphic on the type classes being used then (IMO) there is no benefit. And if the function isn't morally an Alternative (if it breaks if you add more guards) then it actually creates bugs. Direct version simply does what it says.

[1] This current version of Monad and Alternative with Maybe does do the same thing it did before, but it's somewhat confusing because a adding alternative Maybe's returns the last one, while the Monad instance short-circuits, so what does it actually do I had to need to think about :).

Reply:
Let's agree we disagree.
And yes, it's not about being allowed to do things or not, it's about having a codebase we're comfortable with.
As soon as we have two conditionals, we have code like (or even worse (IMO) if we omit the do-notation):

```
if condition1
then do
    x <- someComputation
    if condition2
    then ...
    else Nothing
else Nothing
```

which becomes harder to read for me than

```
do
    guard condition1
    x <- someComputation
    guard $ condition2 x
   ...
```

For consistency, even if there is only one conditional, I'm used to stick to the latter. No sound instance of `Alternative` and `Monad` should be able to escape this behaviour, and I'm not sure we should avoid some idioms to consider unsound instances (and again I didn't want to consider them in the first place).

Reply:
I'm not saying that guard isn't useful in situations where you know that you want this short-circuiting behavior, but as this is a small helper function then I have no idea if we do. And also, now if I really want to make sure that this function does what it was intended to do then I need to go and look up Alternative and Monad for Maybe, whereas previously I could just look at the code and know.